keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md

```diff
@@ -0,0 +1,906 @@
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: PersistentVolumeLastPhaseTransitionTime
+  - Components depending on the feature gate: kube-apiserver, kube-controller-manager
+- [ ] Other
+  - Describe the mechanism:
+  - Will enabling / disabling the feature require downtime of the control
+    plane?
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+###### Does enabling the feature change any default behavior?
+
+<!--
+Any change of default behavior may be surprising to users or break existing
+automations, so be extremely careful here.
+-->
+
+No. It only adds a new informative field to PV status.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+Yes. This will result in the timestamp value being set to `nil`. Mentioned in "Upgrade / Downgrade Strategy" section.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+No issues expected, after rollback the field can be `nil` and validation should allow updates from `nil`.
```

Comment:
I guess I missed that in the proposal, but when exactly this field will be set?

I'm assuming that we the status computed in the controller will be different and we will be patching the object.
If so, what happens after enabling the feature for the first time?
We have no idea when the last transition happened, so what we're going to do:
- will the field remain to be unset?
- are we going to (somewhat misleadingly) set it to now

Reply:
[I guess the former, but then let's also add to Caveats section that the field will remain unset until the first change in PV status even after enabling the feature.]

Reply:
> If so, what happens after enabling the feature for the first time?

If feature is enabled for the first time there are two sub-cases where phase is modified that we need to think of:
  1) While provisioner creates a PV it's phase is set to `Bound` - at this point I'd rather not set the field because it would be misleading, as you said. This field is meant for "last phase change" so it does not make sense to insert something else (like current time) just for the sake of having some value.
  2) Updating volume status (`updateVolumePhase()`) - this is when the new `LastPhaseTransitionTime` field and its value appears in PV object for the first time. So it's good idea to write down the caveeats since enabling FG won't have an immediate effect.

Reply:
OK - so please put that explicitly into your KEP so it's clear what exactly we can expect from this field.

Reply:
Updated this section. Also adding one more design detail to explicitly state that we want to allow timestamp updates, not just from `nil`.